Extensions/Xtensive.Orm.Tracking/Internals/TrackingItem.cs

```diff
@@ -16,23 +16,15 @@ namespace Xtensive.Orm.Tracking
   [DebuggerDisplay("{Key}")]
   internal sealed class TrackingItem : ITrackingItem
   {
-    private IList<ChangedValue> cachedChangedValues;
+    private IReadOnlyList<ChangedValue> cachedChangedValues;
 
-    public Key Key { get; private set; }
+    public Key Key { get; }
 
     public DifferentialTuple RawData { get; private set; }
 
     public TrackingItemState State { get; private set; }
 
-    public IList<ChangedValue> ChangedValues
-    {
-      get
-      {
-        if (cachedChangedValues==null)
-          cachedChangedValues = CalculateChangedValues().ToList().AsReadOnly();
-        return cachedChangedValues;
-      }
-    }
+    public IReadOnlyList<ChangedValue> ChangedValues => cachedChangedValues ??= CalculateChangedValues().ToList().AsReadOnly();
```

Comment:
What are the benefits of this breaking change of API? Is it to make result type more obvious?

If we changed property result type wouldn't it be beneficial to change CalculateChangedValues() like so

```csharp

   private IReadOnlyList<ChangedValue> CalculateChangedValues()
    {
      var originalValues = RawData.Origin;
      var changedValues = RawData.Difference;

      if (State==TrackingItemState.Created) {
        originalValues = null;
        changedValues = RawData.Origin;
      }

      var changedValuesList = new List<ChangedValue>(Key.TypeInfo.Fields.Count);

      foreach (var field in Key.TypeInfo.Fields.Where(f => f.Column != null)) {
        object origValue = null, changedValue = null;
        int fieldIndex = field.MappingInfo.Offset;
        TupleFieldState fieldState;
        if (originalValues != null)
          origValue = originalValues.GetValue(fieldIndex, out fieldState);
        if (changedValues != null) {
          changedValue = changedValues.GetValue(fieldIndex, out fieldState);
          if (!fieldState.IsAvailable())
            continue;
        }
        changedValuesList.Add(new ChangedValue(field, origValue, changedValue));
      }

      return changedValuesList.ToArray();
    }
```

Reply:
`IReadOnlyList<>` result type is more safe than `IList<>`
Changed the suggested way but without `.ToArray()` to save allocation